### PR TITLE
Dependency tree analysis

### PR DESCRIPTION
**Pull request by oganexon**
@divy-work I've moved some files around, tell me how you want to structure the project between Deno's only files, those for the web and the rest.

So I propose the possibility to build a dependency tree and to analyze its imports.
```ts
function dependencyTree(path: string, options: TreeOptions): Promise<IDependencyTree>

//Where :
export type DependencyTree = Array<{
  path: string;
  imports: DependencyTree;
}>;

export interface IDependencyTree {
  tree: DependencyTree; // The actual tree
  circular: boolean; // Are there circular imports?
  count: number; // The number of imported files, by removing duplicate entries
  iterator: IterableIterator<string>; // An iterable dependency iterator to, for example, detect whether the tree contains broken dependencies. 
}

export interface TreeOptions {
  fullTree: boolean; // By default, redundant imports are ignored, i.e. they are already elsewhere in the tree and are not duplicated
}
```

Currently, it only works for Deno but there is almost nothing to change to make it work on the web, tell me how to make it clean without duplicate functions.

```diff
/* Resolves any path, relative or HTTP url. */
export function resolveURL(path: string, base = "") {
  if (path.match(/^https?:\/\//)) {
    return path;
  }
  if (base.match(/^https?:\/\//)) {
    return new URL(path, base).href;
  }
 - return fileURL(path, base);
}

/* Fetch data from file: or https: urls */
async function fetchData(url: string) {
-  if (url.match(/^https?:\/\//)) {
    const data = await fetch(url);
    return data.text();
-  }
-  const data = await Deno.readFile(resolve(fromFileUrl(url)));
-  return decoder.decode(data);
}
```

### CODE REVIEW

File: deno/analyze.ts [hunk 1]
```diff
@@ -2,11 +2,11 @@ import init, {
   source,
   tree as wasm_tree,
   analyze as wasm_analyze,
-} from "./wasm.js";
+} from "../wasm/wasm.js";
 
 await init(source);
 
-export function tree(filename: string, src: string) {
+export function extractDependencies(filename: string, src: string) {
```
Comment (oganexon) [inline deno/analyze.ts:9]: I think the name makes more sense

Comment (divy-work) [reply deno/analyze.ts:9]: `tree` is much shorter tho, let's keep it that way :smile:

File: egg.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "name": "analyzer",
+  "description": "Analyze and detect broken and malicious JavaScript and TypeScript modules.",
+  "files": [
+    "./deno/**/*",
+    "./wasm/wasm.js"
+  ]
+}
```
Comment (divy-work) [inline egg.json:8]:
```suggestion
{
  "name": "analyzer",
  "description": "Analyze and detect broken and malicious JavaScript and TypeScript modules.",
  "stable": true,
  "version": "0.0.1",
  "repository": "https://github.com/nestdotland/analyzer",
  "files": [
    "./wasm.js",
    "./deno.ts"
  ]
}
```

File: deno/dependencyTree.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
+import { extractDependencies } from "./analyze.ts"
+
```
Comment (divy-work) [inline deno/dependencyTree.ts:2]: Let's name this file `tree.ts` instead :)

**Review by divy-work — CHANGES_REQUESTED** [review]

**Review by divy-work — APPROVED** [review]
Yay

**Pull request merged**